File: app/src/main/java/id/ac/stiki/doleno/absenin/view/admin/ui/my_event/MyEventViewModel.java
package id.ac.stiki.doleno.absenin.view.admin.ui.my_event;

import android.os.AsyncTask;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import java.util.List;

import javax.inject.Inject;

import dagger.hilt.android.lifecycle.HiltViewModel;
import id.ac.stiki.doleno.absenin.data.entity.Event;
import id.ac.stiki.doleno.absenin.data.entity.User;
import id.ac.stiki.doleno.absenin.domain.FetchAllEventByEmail;
import id.ac.stiki.doleno.absenin.domain.GetAllEvent;
import id.ac.stiki.doleno.absenin.domain.GetUser;

@HiltViewModel
public class MyEventViewModel extends ViewModel {
    private final GetAllEvent getAllEvent;
    private final GetUser getUser;
    private final FetchAllEventByEmail fetchAllEvent;
    private final MutableLiveData<MyEventState> _myEventState = new MutableLiveData<>();
    public LiveData<MyEventState> myEventState = _myEventState;
    public LiveData<List<Event>> listMyEvent;

    @Inject
    public MyEventViewModel(GetAllEvent getAllEvent, GetUser getUser, FetchAllEventByEmail fetchAllEvent) {
        this.getAllEvent = getAllEvent;
        this.listMyEvent = getAllEvent.execute();
        this.getUser = getUser;
        this.fetchAllEvent = fetchAllEvent;
    }

    public void fetchMyEvent() {
        _myEventState.postValue(MyEventState.LOADING);
        AsyncTask.execute(() -> {
            User user = getUser.execute();
            fetchAllEvent.execute(user.email)
                    .addOnSuccessListener(success -> _myEventState.postValue(MyEventState.SUCCESS))
                    .addOnFailureListener(failure -> _myEventState.postValue(MyEventState.FAILED));
        });
    }

    public Boolean isListEventEmpty() {
        return listMyEvent.getValue().isEmpty();
    }
}

File: app/src/main/java/id/ac/stiki/doleno/absenin/view/participant/ui/my_event/MyEventViewModel.kt
package id.ac.stiki.doleno.absenin.view.participant.ui.my_event

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import dagger.hilt.android.lifecycle.HiltViewModel
import id.ac.stiki.doleno.absenin.data.entity.Absent
import id.ac.stiki.doleno.absenin.domain.FetchAllAbsentHistory
import id.ac.stiki.doleno.absenin.domain.GetAbsentByStatus
import id.ac.stiki.doleno.absenin.domain.GetUser
import id.ac.stiki.doleno.absenin.util.enums.AbsentStatus
import java.util.concurrent.Executors
import javax.inject.Inject

@HiltViewModel
class MyEventViewModel @Inject constructor(
    private val getUser: GetUser,
    private val fetchAllAbsent: FetchAllAbsentHistory,
    private val getAbsentByStatus: GetAbsentByStatus
) : ViewModel() {
    private val _myEventState: MutableLiveData<MyEventState> = MutableLiveData()
    val myEventState: LiveData<MyEventState> = _myEventState
    var absents: List<Absent> = emptyList()

    fun fetchAllAbsent() {
        _myEventState.postValue(MyEventState.LOADING)

        Executors.newSingleThreadExecutor().execute {
            val user = getUser.execute()
            fetchAllAbsent.execute(user.email)
                .addOnSuccessListener {
                    Executors.newSingleThreadExecutor().execute {
                        absents = getAbsentByStatus.execute(listOf(AbsentStatus.REGISTERED.text))
                        _myEventState.postValue(MyEventState.SUCCESS)
                    }
                }
                .addOnFailureListener {
                    _myEventState.postValue(MyEventState.FAILED)
                }
        }
    }

    fun isAbsentsEmpty(): Boolean = absents.isEmpty()
}